src/lib/fetchGuildApi.ts

```diff
@@ -0,0 +1,127 @@
+import { tryGetToken } from "@/actions/token";
+import { env } from "./env";
+import type { ErrorLike } from "./types";
+
+type FetchResult<Data, Error> =
+  | { status: "error"; data: Error; response: Response }
+  | { status: "success"; data: Data; response: Response };
+
+// TODO: include a dedicated logger with severity channels
+const logger = {
+  info: (...args: Parameters<typeof console.info>) => {
+    if (process.env.NODE_ENV === "development" && env.LOGGING > 0) {
+      console.info(
+        `[${new Date().toLocaleTimeString()} - fetchGuildApi]:`,
+        ...args,
+      );
+    }
+  },
+};
+
+/**
+ * Fetcher used for creating requests to the v3 backend API.
+ *
+ * This function sends a request to the specified `pathname` of the API and returns
+ * a structured result containing the response, status, and data. It ensures that the API transmission is with valid JSON.
+ *
+ * Authentication token is retrieved and automatically included in the request headers. The token is retrieved either server-side or client-side
+ * depending on the runtime environment.
+ *
+ * @template Data - The expected shape of the success response data.
+ * @template Error - The expected shape of the error response data.
+ *
+ * @param pathname - The API endpoint to fetch, without leading or trailing slashes.
+ * @param requestInit - Optional configuration for the fetch request.
+ *
+ * @returns A promise resolving to the fetch result,
+ * containing `status`, `data`, and `response`.
+ *
+ * @example
+ * ```ts
+ * const { data, status, response } = await fetchGuildApi<{ name: string }>('guild');
+ *
+ * if (status === 'error') {
+ *   assert(!response.ok);
+ * } else {
+ *   console.log(data.name);
+ * }
+ * ```
+ *
+ * @throws If `pathname` starts or ends with a slash.
+ * @throws If the API does not respond with JSON.
+ * @throws If the response JSON cannot be parsed.
+ */
+export const fetchGuildApi = async <Data = object, Error = ErrorLike>(
+  pathname: string,
+  requestInit?: RequestInit,
+): Promise<FetchResult<Data, Error>> => {
+  if (pathname.startsWith("/")) {
+    throw new Error("`pathname` must not start with slash");
+  }
+  if (pathname.endsWith("/")) {
+    throw new Error("`pathname` must not end with slash");
+  }
+  const url = new URL(`api/${pathname}`, env.NEXT_PUBLIC_API);
+
+  let token: string | undefined;
+  try {
+    token = await tryGetToken();
+  } catch (_) {
```

Comment:
Currently the token error is not propagated and it probably won't be later to keep the auth automatic during fetching.